### PR TITLE
Rebucket Mempool Transactions When Global Params Change

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1231?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1231** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1231?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1230** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feature/proof-of-stake`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tholonious and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/pos_mempool.go [hunk 1]
```diff
@@ -999,6 +999,28 @@ func (mp *PosMempool) pruneNoLock() error {
...
 
+func (mp *PosMempool) rebucketTransactionRegister() error {
+	// Check if the global params haven't changed in a way that requires rebucketing the
+	// transaction register
+	if mp.txnRegister.HasGlobalParamChange(mp.globalParams) {
+		return nil
+	}
+
+	// Rebucket the transaction register
+	newTxnRegister, err := mp.txnRegister.CopyWithNewGlobalParams(mp.globalParams)
```
Comment (lazynina) [inline lib/pos_mempool.go:1010]: do we need to grab a read lock on the `txnRegister` here? Okay just checked, `CopyWithGlobalParams` acquires the lock for us.

File: lib/pos_mempool.go [hunk 2]
```diff
@@ -999,6 +999,28 @@ func (mp *PosMempool) pruneNoLock() error {
 	return nil
 }
 
+func (mp *PosMempool) rebucketTransactionRegister() error {
```
Comment (lazynina) [inline lib/pos_mempool.go:1002]: Can we rename this `rebucketTransactionRegisterNoLock`?

**Review by lazynina — APPROVED** [review]
just one nit on naming.

**Pull request merged**